crud/src/main/java/com/redhat/lightblue/assoc/scorers/IndexedFieldScorer.java

```diff
@@ -123,53 +85,70 @@ public Comparable score(QueryPlan qp) {
             throw new IllegalStateException("Unable to find root metadata");
         }
 
-        CostAndSize rootCost=getAncestorCostAndSize(root);
-        LOGGER.debug("up to root: {}",rootCost);
-        
-        // Evaluation changes after root is retrieved. Any query
-        // evaluation won't affect the result set size, but add to the
-        // cost, because that means we have to manually filter out
-        // results
-
-        BigInteger cost=BigInteger.ONE;
-        for(QueryPlanNode dst:root.getDestinations()) {
-            // If there's a query here, increase the cost based on root size
-            cost=cost.multiply(((IndexedFieldScorerData)dst.getData()).estimatedRootDescendantCost(rootCost.size));
+        BigInteger finalCost;
+        // If the root entity node is the root node, then there is only the cost of retrieval
+        // Otherwise, the cost of retrieval and cost of query must be considered
+        if(sources.length==1&&sources[0]==root) {            
+            // If the root entity has identity search and the query plan
+            // node for the root entity is at the plan root, then this is
+            // the lowest cost option
+            if(((IndexedFieldScorerData)root.getData()).isIdentitySearch() ) {
+                finalCost=BigInteger.ZERO;
+            } else {
+                finalCost=computeRetrievalCost(root);
+            }
+        } else {
+            // Multiple roots, there is both a query and retrieval phase
+            CostAndSize cs=computeQueryCost(root,qp);
+            // Have to retrieve for every result in query
+            finalCost=cs.cost.add(cs.size);
         }
-        // cost is based on size so far, multiple this by root cost for final cost
-        cost=cost.multiply(rootCost.cost);
-        LOGGER.debug("Final cost:{}",cost);
-        return new Score(cost);
-
+        LOGGER.debug("Final cost:{}",finalCost);
+        return finalCost;
     }
 
-    private CostAndSize getAncestorCostAndSize(QueryPlanNode anc) {
-        CostAndSize incoming=null;
-        for(QueryPlanNode src:anc.getSources()) {
-            CostAndSize acs=getAncestorCostAndSize(src);
-            
-            if(incoming==null)
-                incoming=new CostAndSize();
-            
-            // The input result set size is the join size of all incoming nodes
-            incoming.size=incoming.size.multiply( acs.size );
-            // Add up costs
-            incoming.cost=incoming.cost.add( acs.cost );
+    private BigInteger computeRetrievalCost(QueryPlanNode root) {
+        CostAndSize cs=((IndexedFieldScorerData)root.getData()).getCostAndSize();
```

Comment:
What's the problem with it? This class has the factory for QueryPlanData, so it expects them to be of this type.